editor-extensions/vscode/client/src/CodelensProvider.ts

```diff
@@ -8,29 +10,48 @@ import * as vscode from 'vscode';
  */
 export class CodelensProvider implements vscode.CodeLensProvider {
 
+
+
 	private codeLenses: vscode.CodeLens[] = [];
 	private regex: RegExp;
-	private _onDidChangeCodeLenses: vscode.EventEmitter<void> = new vscode.EventEmitter<void>();
-	public readonly onDidChangeCodeLenses: vscode.Event<void> = this._onDidChangeCodeLenses.event;
+	public onDidChangeCodeLensesEmitter: vscode.EventEmitter<void> = new vscode.EventEmitter<void>();
+	public readonly onDidChangeCodeLenses: vscode.Event<void> = this.onDidChangeCodeLensesEmitter.event;
 
-	constructor() {
+	constructor(readonly client: LanguageClient
+	) {
 		this.regex = /(.+)/g;
 
 		vscode.workspace.onDidChangeConfiguration((_) => {
-			this._onDidChangeCodeLenses.fire();
+			this.onDidChangeCodeLensesEmitter.fire();
 		});
 	}
 
-	public provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken): vscode.CodeLens[] | Thenable<vscode.CodeLens[]> {
-		return [];
+	public async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken): Promise<vscode.CodeLens[]> {
+
+
+		const args = { textDocument: { uri: document.uri.toString() } };
+
+		const lenses = await this.client.sendRequest(CodeLensRequest.type, args);
+		return lenses.map(l => {
+			const range = new vscode.Range(
+				new vscode.Position(l.range.start.line, l.range.start.character),
+				new vscode.Position(l.range.end.line, l.range.end.character)
+			);
+
+			const command: vscode.Command = {
+				title: l.data,
+				command: undefined,
```

Comment:
This feels a little weird in general (since in JS you just wouldn't supply this field). Is it a TS thing to have it still provide the key here?

Reply:
Yeah, ugly but it makes TS happy, otherwise you get:

![image](https://user-images.githubusercontent.com/20163757/103155913-99287300-479b-11eb-9dbc-cebaf0bf6bfd.png)

Reply:
Ah, figures.